# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Time & Space Complexity</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a)How do we measure it?<br>
8) <b>In/Out-Place Algorithms</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Out of Place Algorithm <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) In-Class Exercise #1 <br>
9) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>


## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def addTwo(x):
    return x + 2

print(addTwo(4))

# Lambda Function Syntax
# Example of calling a lambda function without a variable
print((lambda x: x + 2)(4))

6
6


#### Saving to a Variable

In [3]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [5]:
# Multiple Inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

# Multiple Inputs with a variable attached
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [6]:
def multiply(f,num):
    """
        f expects a lambda function
        num expects a number
    """
    return f(num)
multiply(lambda x: x * x, 4) # lambda 4: 4 * 4

16

#### Returning a Lambda from a Function

In [18]:
# Regular defined function
def multiply_test(num):
    return num * 4

# Function within a function
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x,a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x00000281F87A3EE8>
8
19


#### If Statements within Lambdas

In [22]:
# Lambda x : True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [23]:
f = lambda cube: cube ** 3

print(f(2))
print(f(3))
print(f(4))

8
27
64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [25]:
# map(func,iterable(list,dict,tuple,set))
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num,num2
    
numbers = [4,11,20,3,15,20]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared,numbers,more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [48]:
# map(lambda x: x + 2, list)
# using lambda in map happens in line (or in one line) usually
# list(map(lambda x : x * 2,nums))
squared_nums_lamb = list(map(lambda x,y : (x ** 2,y ** 2) if x < 10 and y < 10 else (x,y),numbers,more_nums))
print(squared_nums_lamb)


[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [51]:
double_num = list(map(lambda x,y: (((x * 2) - 1),((y * 2) - 1)) if x > 0 and y > 0 else (x,y), numbers,more_nums))
print(double_num)

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [56]:
names = ['Bob','Andy','Max','Evan','Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)

print(a_names)

['Andy', 'Angelica']
<function a_names at 0x00000281F87BF0D8>


#### Using Lambda's with Filter()

In [57]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [62]:
from statistics import mean

nums = [2,7,4.2,1.6,9,4.4,4.9]

print(mean(nums))

avg = mean(nums)
print(avg)

def mean_test(name):
    if name < avg:
        return True
    else:
        return False
new_calc = list(filter(mean_test,nums))
print(new_calc)

# Lambda function solution
below_avg = list(filter(lambda x: True if x <= mean(nums) else False, nums))
print(below_avg)

4.728571428571429
4.728571428571429
[2, 4.2, 1.6, 4.4]
[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [67]:
from functools import reduce

# reduce(function,iterable)

list_1 = [2,4,6,8,10]

def addNums(num1,num2):
    return num1 + num2

result_add = reduce(addNums,list_1)

print(result_add)

# Subtract a list of numbers
def subtractNums(num1,num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)


30
-26


#### Using Lambda's with Reduce()

In [68]:
result_lamb = reduce(lambda x,y: x + y, list_1)
print(result_lamb)

30


In [69]:
my_list = [1,2,3,4]

from functools import reduce

result_lamb = reduce(lambda x,y: x * y, my_list)
print(result_lamb)


24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

#### Implementing a Base Case

In [1]:
def addNums(num):
    # Set base case for recusive function
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num}) = {num} + addNums({num - 1})")
        return num + addNums(num - 1)
    
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [2]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
    
factorial(5)


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [1]:
# return num - subNums(num - 1)
# subtract(10)
def subNums(num):
    if num <= 1:
        print("subNums(1) = 1")
        return num
    else:
        print(f"subNums({num}) = {num} - subNums({num - 1})")
        return num - subNums(num - 1)
    
subNums(5)

subNums(5) = 5 - subNums(4)
subNums(4) = 4 - subNums(3)
subNums(3) = 3 - subNums(2)
subNums(2) = 2 - subNums(1)
subNums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [14]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

my_range(20,start = 2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x000002EB1C6E5C48>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [17]:
# for i in numbers_squared(10)

def num_sq(stop,start,step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in num_sq(0,start = 3):
    print(i)

9
4
1
0


## Time and Space Complexity

#### What is it?

Time and space complexity is the measure of how much time a given action(function) will take to solve a problem. In the same fashion, we determine how much a given data structure will need in terms of memory allocation. A problem can have multiple solutions and finding the optimal solution for the problem needs to be analyzed in time and space.

#### How do we measure Time and Space Complexity?

In order to measure time and space complexity we use Asymptotic analysis. The reason for this is because we need a way to measure different algorithms (functions) based on the size of their inputs in a mathmatical way. For example, we could have a function that is computed as f(n) and another that is g(n^2). All things around the function staying constant, the only thing that changes is the size of the input. Below is the chart that shows the different Asymptotic analysis formats. 

<table style="text-align:center;" class="table table-bordered">
<tbody><tr>
<td>constant</td>
<td>−</td>
<td>Ο(1)</td>
</tr>
<tr>
<td>logarithmic</td>
<td>−</td>
<td>Ο(log n)</td>
</tr>
<tr>
<td>linear</td>
<td>−</td>
<td>Ο(n)</td>
</tr>
<tr>
<td>Linear Logarithmic</td>
<td>−</td>
<td>Ο(n log n)</td>
</tr>
<tr>
<td>quadratic</td>
<td>−</td>
<td>Ο(n<sup>2</sup>)</td>
</tr>
<tr>
<td>cubic</td>
<td>−</td>
<td>Ο(n<sup>3</sup>)</td>
</tr>
<tr>
<td>polynomial</td>
<td>−</td>
<td>n<sup>Ο(1)</sup></td>
</tr>
<tr>
<td>exponential</td>
<td>−</td>
<td>2<sup>Ο(n)</sup></td>
</tr>
</tbody></table>

## In-Place Algorithms

#### Syntax

In [ ]:
# var[i],var[i + 1] = var[i + 1], var[i]
# Sometimes known as a swap algorithm
def swap(alist,x,y,z):
    alist[x],alist[y],alist[z] = alist[z],alist[y],alist[x]
    return alist

my_list = [20,4,10]
print(f"Before swap: {my_list}")

# swap(my_list,0,1,2)
swap(my_list,2,0,1)


print(f"AFTER swap: {my_list}")

#### Out of Place Algorithm

In [ ]:
# not swapping, but rather reversing completely
# BUT also copies to another place in memory
my_list_copy = my_list[::-1]
print(my_list_copy)

array = ['a', 'b', 'c', 'd']
new_array = ['a'] * len(array)

print("Before",array)

length = len(array) - 1

for i in range(length):
    new_array[i] = array[length - i]
    
array = new_array
print("After",array)


#### In-Class Exercise <br>
<p>Write a function that takes in four arguments (list, index1, index2, index3), and swaps those three positions in the list passed in.</p>

In [ ]:
l_1 = [10, 4, 3, 8, 4, 2, 6]

def switch(lists,x,y,z):
    lists[x],lists[y],lists[z] = lists[z],lists[x],lists[y]
    return lists

#solution 2
def switch(lists,a,b,c,d,e):
    lists[a],lists[b],lists[c], lists[d],lists[e] = lists[e],lists[d],lists[c],lists[b],lists[a]
    return lists

print(switch(l_1,0,1,2))

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [5]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def empty_string_search(string):
    if string.isspace() == True or string == '':
        return False
    else:
        return True

filtered_list = list(filter(empty_string_search,places))

print(filtered_list)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [8]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

# def make_dict(name):
#     return name.title().split()
        

# names_list = list(map(make_dict, author))
# print(names_list)

# output = list(map(lambda name: name.split().sort(key=), author))

# print(output)


for peeps in author:
    peeps.lower().split()
    print(peeps.lower().split()[-1])

author.sort(key=lambda name: name.title().split()[-1])

print(author)
# print(ord("C"))


carter
anisimov
garfield
hasselhoff
bernstein
['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [14]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

to_farhenheit = list(map(lambda place: (9/5)*place[1] +32, places))

print(to_farhenheit)

[89.6, 53.6, 111.2, 84.2]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [6]:
def fib(num):
   if num == 1: #base case, checks to see if we have reached the starting point of our Fibonacci sequence (1)
      return 1
   elif num == 2:##Since each number in the sequence is equal to the sum of the two previous numbers, 
      return 1   #the second iteration should equal 1. However,I don't understand why we need to specify this, 
   else:         #as fib(2 - 1) + fib(2 - 2) already equals 1, but the code  gets stuck in an infinite loop without it. ## 
      return fib(num -1) + fib(num -2)    #Perhaps because there is only one iteration to reference before that and it needs two to evaluat correctly? 
            #^This statement simply sums the two previous numbers. 

for i in range(1,30): #My function only returns one number at a time, so it needs to be repeated on every number in the range. Or maybe mapped?
   print(f"iteration {i - 1}: {fib(i)}")     #I will experiment with that in my free time.
    


iteration 0: 1
iteration 1: 1
iteration 2: 2
iteration 3: 3
iteration 4: 5
iteration 5: 8
iteration 6: 13
iteration 7: 21
iteration 8: 34
iteration 9: 55
iteration 10: 89
iteration 11: 144
iteration 12: 233
iteration 13: 377
iteration 14: 610
iteration 15: 987
iteration 16: 1597
iteration 17: 2584
iteration 18: 4181
iteration 19: 6765
iteration 20: 10946
iteration 21: 17711
iteration 22: 28657
iteration 23: 46368
iteration 24: 75025
iteration 25: 121393
iteration 26: 196418
iteration 27: 317811
iteration 28: 514229
